### PR TITLE
boards: riscv: initial support of Andes FPGA-based RISC-V platform adp_xc7k_ae350

### PR DESCRIPTION
**Pull request by cwshu**
`adp_xc7k_ae350` board is for [AE350 RISC-V platform](http://www.andestech.com/en/products-solutions/andeshape-platforms/ae350-axi-based-platform-pre-integrated-with-n25f-nx25f-a25-ax25/) on [ADP-XC7K series FPGA-based development boards](http://www.andestech.com/en/products-solutions/andeshape-platforms/adp-xc7k160-410/). This board supports generic [V5 AndesCore](http://www.andestech.com/en/products-solutions/andescore-processors/) RISC-V processors (e.g. 25/27/45-series) on AE350 platform.

This initial patch contains SoC, board, and UART/GPIO drivers. Also, the SoC porting contains Andes V5 CPU-specific features including our custom CSRs, Andes V5 PMA (support nocache memory), and L2 cache controller.

In our future plan, the complete support of `adp_xc7k_ae350` board will contain 8 native Zephyr drivers and RISC-V SMP support (for 25/45MP processor). We already have a downstream version of it [1] in Zephyr v2.4, and also want to upstream these features after we upgrade them to latest Zephyr kernel.

Any suggestion / opinion about our PR is very welcome.

[1] 
Zephyr drivers support: https://github.com/cwshu/zephyr/tree/andes_v5_demo/
RISC-V SMP support: [#29131]

### CODE REVIEW

**Review by katsuster — APPROVED** [review]

**Comment by cwshu** [comment]
Rebase to fix merge conflict of gpio Kconfig/CMakeList.txt because new gpio drivers are added to main

File: soc/riscv/riscv-privilege/andes_v5/Kconfig.defconfig.ae350
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/andes_v5/Kconfig.defconfig.ae350:11]: Why does this depend on `!CACHE_ENABLE`? From what I've checked, the value will default to 100 anyway if `TICKLESS_KERNEL` is set to `n`, and to 10000 otherwise.

File: drivers/gpio/gpio_andes_atcgpio100.c
Comment (fkokosinski) [inline drivers/gpio/gpio_andes_atcgpio100.c:128]: Can a single pin be set to both `INIT_HIGH` and `INIT_LOW`? If not, this probably should be in a if/else if/else block.

File: drivers/gpio/gpio_andes_atcgpio100.c
Comment (fkokosinski) [inline drivers/gpio/gpio_andes_atcgpio100.c:81]: I know that `kernel.h` includes `stdbool.h`, but I've seen other files including it separately when using `false`/`true` values, so to keep consistency you probably should do it as well.

File: drivers/gpio/gpio_andes_atcgpio100.c
Comment (fkokosinski) [inline drivers/gpio/gpio_andes_atcgpio100.c:324]: You could use here (and anywhere else where you use bitmask) the `BIT_MASK` macro from `include/sys/util_macro.h`.

File: boards/riscv/adp_xc7k_ae350/doc/index.rst
Comment (cwshu) [reply boards/riscv/adp_xc7k_ae350/doc/index.rst]: Thanks. This table doesn't need indentation.
Fix it.

File: drivers/gpio/gpio_andes_atcgpio100.c
Comment (cwshu) [reply drivers/gpio/gpio_andes_atcgpio100.c:128]: They can't be set at same time. Using else if will make it more clearly.
Done.

File: drivers/gpio/gpio_andes_atcgpio100.c
Comment (cwshu) [reply drivers/gpio/gpio_andes_atcgpio100.c:82]: Done.

File: drivers/gpio/gpio_andes_atcgpio100.c
Comment (cwshu) [reply drivers/gpio/gpio_andes_atcgpio100.c]: Rebased to BIT / BIT_MASK macro in Zephyr.

But there is a 1 problem. `BIT_MASK(32)` will have integer overflow in `BIT(32)` and get compiler warning (-Wshift-count-overflow).
I have used `BIT64_MASK` and converted to `uint32_t` to workaround it. 
(Compiler could optimize it, but it doesn't seem to be clear?)

File: soc/riscv/riscv-privilege/andes_v5/Kconfig.defconfig.ae350
Comment (cwshu) [reply soc/riscv/riscv-privilege/andes_v5/Kconfig.defconfig.ae350:11]: Because we recommand 100 tick rate in cache disabled mode no matter it's ticked or tickless kernel.

Disabling cache affects system performance much. It can break few kernel testcase (e.g. `tests/kernel/sched/schedule_api`) in the high tick rate because the checking of these cases is based on the tick unit. We'd still like to keep all kernel testcases passed when disabling cache.

The reason why we provide cache disabled mode is that it's helpful to reduce the system complexity of cache (especially DMA coherence) in the development stage or debugging.

**Comment by cwshu** [comment]
Also updates linker script to fix build error of new testcase `tests/misc/iterable_sections`.

File: drivers/gpio/gpio_andes_atcgpio100.c
Comment (cwshu) [reply drivers/gpio/gpio_andes_atcgpio100.c]: I think using `(uint32_t) BIT64_MASK(32)` is OK in this PR, so I will resolve this conversation first. 
If having any problem, you can reopen it.

File: soc/riscv/riscv-privilege/andes_v5/Kconfig.defconfig.ae350
Comment (cwshu) [reply soc/riscv/riscv-privilege/andes_v5/Kconfig.defconfig.ae350:11]: If there isn't any problem, I will resolve this comment.

Comment (fkokosinski) [reply soc/riscv/riscv-privilege/andes_v5/Kconfig.defconfig.ae350:11]: Thanks for the explanation, makes sense. You can resolve this comment. :)

**Comment by cwshu** [comment]
Update RISC-V GP support of andes_v5 soc after #37763 merged

**Comment by cwshu** [comment]
Zephyr v2.7 will close window at one week. This PR still needs 1 approval review. Could anyone help to review our PR?
@kgugala @pgielda @galak @carlescufi

If there isn't enough time for reviewing before v2.7, I can reduce this PR to minimal board support like removing SoC features, gpio driver, or any code can be split to another PR. Maybe it can help some about the problem.

**Review by kgugala — APPROVED** [review]
LGTM

**Pull request merged**
